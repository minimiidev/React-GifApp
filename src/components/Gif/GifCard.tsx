
interface Props {
    id?: string;
    title?: string;
    url?: string;
}

export const GifCard = ({ id, title, url }: Props) => {
    return (
        <div id={id}>
            <img src={url} alt={title} />

            <h4>{title}</h4>
        </div>
    )
}
