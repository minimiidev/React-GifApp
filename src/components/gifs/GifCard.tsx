import { Icon } from '@iconify/react';
import { useRef } from 'react';

interface Props {
    id: string;
    title: string;
    url: string;
}


export const GifCard = ({ id, title, url }: Props) => {
    const imgRef = useRef<HTMLImageElement>(null);

    // TODO: ATOMIZAR ??
    const downloadGif = () => {

        const gif = imgRef.current;
        if (!gif) return;

        const a = document.createElement('a');
        a.href = gif.src;
        a.click();
    }


    return (
        <div id={id} className="relative flex flex-col shadow-md border border-slate-300 w-full hover:scale-95 transition-all duration-300 ease-in-out cursor-pointer">
            <Icon icon="material-symbols:download-sharp" width="24" height="24" className="top-2 right-2 absolute text-white hover:text-blue-500"
                onClick={downloadGif}
            />

            <img ref={imgRef} src={url} alt={title} className="w-full h-full min-h-[200px] object-cover" />

            <h4 className="bg-white p-2 font-medium uppercase">{title}</h4>
        </div>
    )
}


