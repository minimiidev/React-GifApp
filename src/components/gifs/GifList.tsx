import { useState } from "react";
import { useFetch } from "../../hooks";
import { GifCard } from "./GifCard"
import { Icon } from "@iconify/react";

interface Props {
    category: string;
}

export const GifList = ({ category }: Props) => {
    const [favorite, setFavorite] = useState<boolean>(false)

    const { error, images, loading } = useFetch(category)

    return (
        <section className="flex flex-col py-3 w-full">

            <div className="flex justify-between items-center mb-5 pb-2 border-gray-400 border-b">
                <h4 className="font-light text-2xl uppercase">{category}</h4>

                {/* TODO: BOTON DE FAVORITOS */}
                <div className="flex gap-2">
                    <div onClick={() => setFavorite(!favorite)}>
                        {favorite ?
                            <Icon icon="uil:favorite" width="24" height="24" className="text-yellow-500" />
                            :
                            <Icon icon="uim:favorite" width="24" height="24" className="text-yellow-500" />}
                    </div>

                    <Icon icon="oui:save" width="24" height="24" className="text-green-500" />
                    <Icon icon="mi:delete" width="24" height="24" className="text-red-500" />
                </div>
            </div>

            {loading && <h2 className="bg-sky-500 py-10 text-white uppercase">Loading...</h2>}

            {error && <h2>{error.message}</h2>}

            <div className="gap-4 grid grid-cols-3 lg:grid-cols-5">

                {
                    images.map(({ id, title, url }) => (
                        <GifCard key={id} id={id} title={title} url={url} />
                    ))
                }
            </div>
        </section>
    )
}

