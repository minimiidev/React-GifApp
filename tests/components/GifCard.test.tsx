import { render, screen } from '@testing-library/react'
import { GifCard } from '../../src/components/gifs/GifCard';

const id = "1";
const title = "Test gif";
const url = "https://test.com/gif.gif";

describe('Pruebas en <GifCard />', () => {
    test('Debe hacer match con el snapshot', () => {
        const { container } = render(<GifCard id={id} title={title} url={url} />);

        expect(container).toMatchSnapshot();
    })

    test("Debe de mostrar img con URL y ALT indicado", () => {
        render(<GifCard id={id} title={title} url={url} />);
        // screen.debug();
        // expect((screen.getByRole('img') as HTMLImageElement).src).toBe(url);
        // expect((screen.getByRole('img') as HTMLImageElement).alt).toBe(title);
        const { src, alt } = screen.getByRole('img') as HTMLImageElement;
        expect(src).toBe(url);
        expect(alt).toBe(title);
    })

    test("Debe de mostrar el titulo en el componente", () => {
        render(<GifCard id={id} title={title} url={url} />);

        expect(screen.getByText(title)).toBeTruthy();
    })

})