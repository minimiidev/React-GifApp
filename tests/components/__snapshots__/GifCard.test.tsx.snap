// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Pruebas en <GifCard /> Debe hacer match con el snapshot 1`] = `
<div>
  <div
    class="relative flex flex-col shadow-md border border-slate-300 w-full hover:scale-95 transition-all duration-300 ease-in-out cursor-pointer"
    id="1"
  >
    <span />
    <img
      alt="Test gif"
      class="w-full h-full min-h-[200px] object-cover"
      src="https://test.com/gif.gif"
    />
    <h4
      class="bg-white p-2 font-medium uppercase"
    >
      Test gif
    </h4>
  </div>
</div>
`;
